ClassLibrary1/AugmentingGenerator.cs

```diff
@@ -0,0 +1,77 @@
+﻿using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Generators
+{
+    [Generator]
+    public class AugmentingGenerator : ISourceGenerator
+    {
+        public void Initialize(GeneratorInitializationContext context)
+        {
+            // Register a factory that can create our custom syntax receiver
+            context.RegisterForSyntaxNotifications(() => new MySyntaxReceiver());
+        }
+
+        public void Execute(GeneratorExecutionContext context)
+        {
+            // the generator infrastructure will create a receiver and populate it
+            // we can retrieve the populated instance via the context
+            MySyntaxReceiver syntaxReceiver = (MySyntaxReceiver)context.SyntaxReceiver;
+            // get the recorded user class
+            ClassDeclarationSyntax customOptions = syntaxReceiver.ClassToAugment;
+            if (customOptions is null)
+            {
+                // if we didn't find the user class, there is nothing to do
+                return;
+            }
+
+            // add the generated implementation to the compilation
+            SourceText sourceText = SourceText.From($@"
+using System;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+
+namespace CustomOptionsNamespace
+{{
+    public static class CustomOptionsConfigurationExensions
+    {{
+        public static void Bind<TOptions>(this IConfiguration configuration, string key, TOptions options)
+            where TOptions : {customOptions.Identifier}
+        {{
+            configuration.GetSection(key).Bind(options);
+        }}
+
+        public static void Bind<TOptions>(this IConfiguration configuration, TOptions options)
+            where TOptions : {customOptions.Identifier}
+        {{
+            // Generated code
+            options.CustomProperty = ""uncomment the following!"";//configuration[nameof(options.CustomProperty)];
+        }}
+    }}
+}}
+
+", Encoding.UTF8);
+            context.AddSource("CustomOptions.Generated.cs", sourceText);
+        }
+
+        class MySyntaxReceiver : ISyntaxReceiver
+        {
+            public ClassDeclarationSyntax ClassToAugment { get; private set; }
+
+            public void OnVisitSyntaxNode(SyntaxNode syntaxNode)
+            {
+                // Business logic to decide what we're interested in goes here
+                if (syntaxNode is ClassDeclarationSyntax cds &&
+                    cds.Identifier.ValueText == "CustomOptions")
```

Comment:
@davidfowl 's comment https://github.com/maryamariyan/configuration-binding-source-generation/pull/1#discussion_r501299721

We want to look for calls to Bind and Configure instead of looking for CustomOptions.

Reply:
Makes sense. not sure yet what would replace with CustomOptions here.

Reply:
Perhaps we could parse and look for a pattern like "Bind<T>" and retrieve the value T from it to get T? in this case CustomOptions. Just a thought.

Reply:
Or if there was an attribute that could be set on classes like CustomOptions then we could loop through all and generate such code for them

Reply:
> Perhaps we could parse and look for a pattern like "Bind<T>" and retrieve the value T from it and for any such T we find autogenerate the code blob in this file? in this case CustomOptions. Just a thought.

This is what I think we should try.

> Or if there was an attribute that could be set on classes like CustomOptions then we could loop through all and generate such code for them

This works but you'd also need the attribute to have the ability to point to types that you don't own:

```
[ConfigurationBinder(typeof(CustomOptions))]
```

Reply:
> Perhaps we could parse and look for a pattern like "Bind" and retrieve the value T from it and for any such T we find autogenerate the code blob in this file? in this case CustomOptions. Just a thought.

I also think we should pursue this. Also, something to consider, there are some non-generic `Bind` methods on `IConfiguration`, which takes the instance as a parameter.

So I think we should inspect `.Bind` calls and either use the `T` or the instance type if the call is to the non-generic method.

Reply:
@safern I have an issue that I'm almost done filing that says we should add a generic Bind\<T\> overload for the sake of this implementation 😄